// @parcel/utils/lib/replaceBundleReferences.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.replaceURLReferences = replaceURLReferences;
exports.replaceInlineReferences = replaceInlineReferences;

var _stream = require("stream");

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _url = _interopRequireDefault(require("url"));

var _ = require("../");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Replaces references to dependency ids for URL dependencies with:
 *   - in the case of an unresolvable url dependency, the original moduleSpecifier.
 *     These are external requests that Parcel did not bundle.
 *   - in the case of a reference to another bundle, the relative url to that
 *     bundle from the current bundle.
 */
function replaceURLReferences({
  bundle,
  bundleGraph,
  contents,
  map,
  relative = true
}) {
  let replacements = new Map();

  for (let dependency of bundleGraph.getExternalDependencies(bundle)) {
    if (!dependency.isURL) {
      continue;
    }

    let bundleGroup = bundleGraph.resolveExternalDependency(dependency);

    if (bundleGroup == null) {
      replacements.set(dependency.id, {
        from: dependency.id,
        to: dependency.moduleSpecifier
      });
      continue;
    }

    let [entryBundle] = bundleGraph.getBundlesInBundleGroup(bundleGroup);

    if (entryBundle.isInline) {
      // If a bundle is inline, it should be replaced with inline contents,
      // not a URL.
      continue;
    }

    replacements.set(dependency.id, getURLReplacement({
      dependency,
      fromBundle: bundle,
      toBundle: entryBundle,
      relative
    }));
  }

  return performReplacement(replacements, contents, map);
}
/*
 * Replaces references to dependency ids for inline bundles with the packaged
 * contents of that bundle.
 */


async function replaceInlineReferences({
  bundle,
  bundleGraph,
  contents,
  map,
  getInlineReplacement,
  getInlineBundleContents
}) {
  let replacements = new Map();

  for (let dependency of bundleGraph.getExternalDependencies(bundle)) {
    let bundleGroup = bundleGraph.resolveExternalDependency(dependency);

    if (bundleGroup == null) {
      continue;
    }

    let [entryBundle] = bundleGraph.getBundlesInBundleGroup(bundleGroup);

    if (!entryBundle.isInline) {
      continue;
    }

    let packagedBundle = await getInlineBundleContents(entryBundle, bundleGraph);
    let packagedContents = (packagedBundle.contents instanceof _stream.Readable ? await (0, _.bufferStream)(packagedBundle.contents) : packagedBundle.contents).toString();
    let inlineType = (0, _nullthrows.default)(entryBundle.getMainEntry()).meta.inlineType;

    if (inlineType == null || inlineType === 'string') {
      replacements.set(dependency.id, getInlineReplacement(dependency, inlineType, packagedContents));
    }
  }

  return performReplacement(replacements, contents, map);
}

function getURLReplacement({
  dependency,
  fromBundle,
  toBundle,
  relative
}) {
  let url = _url.default.parse(dependency.moduleSpecifier);

  let to;

  if (relative) {
    url.pathname = (0, _.relativeBundlePath)(fromBundle, toBundle, {
      leadingDotSlash: false
    });
    to = _url.default.format(url);
  } else {
    url.pathname = (0, _nullthrows.default)(toBundle.name);
    to = (0, _.urlJoin)((0, _nullthrows.default)(toBundle.target.publicUrl), _url.default.format(url));
  }

  return {
    from: dependency.id,
    to
  };
}

function performReplacement(replacements, contents, map) {
  let finalContents = contents;

  for (let {
    from,
    to
  } of replacements.values()) {
    // Perform replacement
    finalContents = finalContents.split(from).join(to);
  }

  return {
    contents: finalContents,
    // TODO: Update sourcemap with adjusted contents
    map
  };
}
