// @parcel/transformer-js/lib/JSTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _semver = _interopRequireDefault(require("semver"));

var _generator = _interopRequireDefault(require("@babel/generator"));

var _plugin = require("@parcel/plugin");

var _dependencies = _interopRequireDefault(require("./visitors/dependencies"));

var _process = _interopRequireDefault(require("./visitors/process"));

var _fs = _interopRequireDefault(require("./visitors/fs"));

var _globals = _interopRequireDefault(require("./visitors/globals"));

var _parser = require("@babel/parser");

var _traverse = _interopRequireDefault(require("@babel/traverse"));

var walk = _interopRequireWildcard(require("babylon-walk"));

var babelCore = _interopRequireWildcard(require("@babel/core"));

var _scopeHoisting = require("@parcel/scope-hoisting");

var _utils = require("@parcel/utils");

var _sourceMap = _interopRequireDefault(require("@parcel/source-map"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const IMPORT_RE = /\b(?:import\b|export\b|require\s*\()/;
const ENV_RE = /\b(?:process\.env)\b/;
const BROWSER_RE = /\b(?:process\.browser)\b/;
const GLOBAL_RE = /\b(?:process|__dirname|__filename|global|Buffer|define)\b/;
const FS_RE = /\breadFileSync\b/;
const SW_RE = /\bnavigator\s*\.\s*serviceWorker\s*\.\s*register\s*\(/;
const WORKER_RE = /\bnew\s*(?:Shared)?Worker\s*\(/; // Sourcemap extraction
// const SOURCEMAP_RE = /\/\/\s*[@#]\s*sourceMappingURL\s*=\s*([^\s]+)/;
// const DATA_URL_RE = /^data:[^;]+(?:;charset=[^;]+)?;base64,(.*)/;

function canHaveDependencies(code) {
  return IMPORT_RE.test(code) || GLOBAL_RE.test(code) || SW_RE.test(code) || WORKER_RE.test(code);
}

var _default = new _plugin.Transformer({
  canReuseAST({
    ast
  }) {
    return ast.type === 'babel' && _semver.default.satisfies(ast.version, '^7.0.0');
  },

  async parse({
    asset,
    options
  }) {
    let code = await asset.getCode();

    if (!options.scopeHoist && !canHaveDependencies(code) && !ENV_RE.test(code) && !BROWSER_RE.test(code) && !FS_RE.test(code)) {
      return null;
    }

    return {
      type: 'babel',
      version: '7.0.0',
      isDirty: false,
      program: (0, _parser.parse)(code, {
        filename: this.name,
        allowReturnOutsideFunction: true,
        strictMode: false,
        sourceType: 'module',
        plugins: ['exportDefaultFrom', 'exportNamespaceFrom', 'dynamicImport']
      })
    };
  },

  async transform({
    asset,
    options,
    logger
  }) {
    asset.type = 'js';

    if (!asset.ast) {
      return [asset];
    }

    let ast = asset.ast;
    let code = await asset.getCode(); // Inline process/ environment variables

    if (!asset.env.isNode() && (ast.isDirty || ENV_RE.test(code)) || asset.env.isBrowser() && (ast.isDirty || BROWSER_RE.test(code))) {
      walk.ancestor(ast.program, _process.default, {
        asset,
        env: options.env,
        isNode: asset.env.isNode(),
        isBrowser: asset.env.isBrowser()
      });
    } // Collect dependencies


    if (canHaveDependencies(code) || ast.isDirty) {
      walk.ancestor(ast.program, _dependencies.default, {
        asset,
        options
      });
    } // If there's a hashbang, remove it and store it on the asset meta.
    // During packaging, if this is the entry asset, it will be prepended to the
    // packaged output.


    if (ast.program.program.interpreter != null) {
      asset.meta.interpreter = ast.program.program.interpreter.value;
      delete ast.program.program.interpreter;
    }

    if (!asset.env.isNode()) {
      // Inline fs calls
      let fsDep = asset.getDependencies().find(dep => dep.moduleSpecifier === 'fs');

      if (fsDep && FS_RE.test(code)) {
        // Check if we should ignore fs calls
        // See https://github.com/defunctzombie/node-browser-resolve#skip
        let pkg = await asset.getPackage();
        let ignore = pkg && pkg.browser && typeof pkg.browser === 'object' && pkg.browser.fs === false;

        if (!ignore) {
          (0, _traverse.default)(ast.program, _fs.default, null, {
            asset,
            logger
          });
        }
      } // Insert node globals


      if (GLOBAL_RE.test(code)) {
        asset.meta.globals = new Map();
        walk.ancestor(ast.program, _globals.default, asset);
      }
    }

    if (options.scopeHoist) {
      (0, _scopeHoisting.hoist)(asset);
    } else if (asset.meta.isES6Module) {
      // Convert ES6 modules to CommonJS
      let res = babelCore.transformFromAst(ast.program, code, {
        code: false,
        ast: true,
        filename: asset.filePath,
        babelrc: false,
        configFile: false,
        plugins: [require('@babel/plugin-transform-modules-commonjs')]
      });
      ast.program = res.ast;
      ast.isDirty = true;
    }

    return [asset];
  },

  async generate({
    asset,
    options
  }) {
    let code = await asset.getCode();
    let res = {
      code
    };
    let ast = asset.ast;

    if (ast && ast.isDirty !== false) {
      let sourceFileName = (0, _utils.relativeUrl)(options.projectRoot, asset.filePath);
      let generated = (0, _generator.default)(ast.program, {
        sourceMaps: options.sourceMaps,
        sourceFileName: sourceFileName
      }, code);
      res.code = generated.code; // $FlowFixMe...

      res.map = new _sourceMap.default(generated.rawMappings, {
        [sourceFileName]: null
      });
    }

    if (asset.meta.globals && asset.meta.globals.size > 0) {
      res.code = Array.from(asset.meta.globals.values()).map(g => g ? g.code : '').join('\n') + '\n' + res.code;
    }

    delete asset.meta.globals;
    return res;
  }

});

exports.default = _default;
