// @parcel/packager-css/lib/CSSPackager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _plugin = require("@parcel/plugin");

var _utils = require("@parcel/utils");

var _default = new _plugin.Packager({
  async package({
    bundle,
    bundleGraph,
    getInlineBundleContents
  }) {
    let queue = new _utils.PromiseQueue({
      maxConcurrent: 32
    });
    bundle.traverseAssets({
      exit: asset => {
        // Figure out which media types this asset was imported with.
        // We only want to import the asset once, so group them all together.
        let media = [];

        for (let dep of bundleGraph.getIncomingDependencies(asset)) {
          if (!dep.meta.media) {
            // Asset was imported without a media type. Don't wrap in @media.
            media.length = 0;
            break;
          }

          media.push(dep.meta.media);
        }

        queue.add(() => asset.getCode().then(css => {
          if (media.length) {
            return `@media ${media.join(', ')} {\n${css.trim()}\n}\n`;
          }

          return css;
        }));
      }
    });
    let outputs = await queue.run();
    return (0, _utils.replaceInlineReferences)({
      bundle,
      bundleGraph,
      // $FlowFixMe
      contents: (0, _utils.replaceURLReferences)({
        bundle,
        bundleGraph,
        contents: outputs.map(output => output).join('\n')
      }).contents,
      getInlineBundleContents,
      getInlineReplacement: (dep, inlineType, contents) => ({
        from: dep.id,
        to: contents
      })
    });
  }

});

exports.default = _default;
